examples/TestEm3/electrons.cu

```diff
@@ -258,11 +265,83 @@ void TransportElectrons(Track *electrons, const adept::MParray *active,
     // numbers after MSC used up a fair share for sampling the displacement.
     currentTrack.rngState.Advance();
 
+    if constexpr (IsElectron) g_nextInteractionForEl[i] = winnerProcessIndex;
+    else g_nextInteractionForPos[i] = winnerProcessIndex;
+    assert(i < (IsElectron ?
+        sizeof(g_nextInteractionForEl)/sizeof(g_nextInteractionForEl[0])
+      : sizeof(g_nextInteractionForPos)/sizeof(g_nextInteractionForPos[0]) ));
+  }
+}
+
+
+// Instantiate kernels for electrons and positrons.
+__global__ __launch_bounds__(ThreadsPerBlock, MinBlocksPerSM)
+void TransportElectrons(Track *electrons, const adept::MParray *active, Secondaries secondaries,
+                        adept::MParray *activeQueue, GlobalScoring *globalScoring,
+                        ScoringPerVolume *scoringPerVolume)
+{
+  TransportElectrons</*IsElectron*/ true>(electrons, active, secondaries, activeQueue, globalScoring, scoringPerVolume);
+}
+__global__ __launch_bounds__(ThreadsPerBlock, MinBlocksPerSM)
+void TransportPositrons(Track *positrons, const adept::MParray *active, Secondaries secondaries,
+                        adept::MParray *activeQueue, GlobalScoring *globalScoring,
+                        ScoringPerVolume *scoringPerVolume)
+{
+  TransportElectrons</*IsElectron*/ false>(positrons, active, secondaries, activeQueue, globalScoring,
+                                           scoringPerVolume);
+}
+
+template<bool IsElectron, int ProcessIndex>
+__global__
+void ComputeInteraction(Track *electrons, const adept::MParray *active, Secondaries secondaries,
+                        adept::MParray *activeQueue, GlobalScoring *globalScoring,
+                        ScoringPerVolume *scoringPerVolume)
+{
+  constexpr unsigned int sharedSize = 12250;
+  __shared__ int candidates[sharedSize];
+  __shared__ unsigned int counter;
+  __shared__ unsigned int noopCounter;
+  counter = 0;
+  noopCounter = 0;
+
+  for (int i = threadIdx.x; i < sharedSize; i += blockDim.x) {
+    candidates[i] = -1;
+  }
+  __syncthreads();
+
+  const int activeSize = active->size();
+  assert(activeSize < sizeof(g_nextInteractionForEl)/sizeof(g_nextInteractionForEl[0]));
+  for (int i = blockIdx.x * blockDim.x + threadIdx.x; i < activeSize; i += blockDim.x * gridDim.x) {
+    const auto winnerProcess = IsElectron ? g_nextInteractionForEl[i] : g_nextInteractionForPos[i];
+
+    if (winnerProcess == ProcessIndex) {
+      const auto destination = atomicInc(&counter, (unsigned int)-1);
```

Comment:
I find `atomicAdd(&counter, 1)` much easier to understand. For `atomicInc`, you need to know that the second argument is the "maximum" where the atomic increment is clamped to...